bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -279,6 +262,59 @@ func (r *KubeadmConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	return r.joinWorker(ctx, scope)
 }
 
+func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, config *bootstrapv1.KubeadmConfig, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	token := config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token
+
+	remoteClient, err := r.remoteClientGetter(ctx, r.Client, util.ObjectKey(cluster))
+	if err != nil {
+		log.Error(err, "Error creating remote cluster client")
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Refreshing token until the infrastructure has a chance to consume it")
+	err = refreshToken(ctx, remoteClient, token)
+	if err != nil {
+		// It would be nice to re-create the bootstrap token if the error was "not found", but we have no way to update the Machine's bootstrap data
+		return ctrl.Result{}, errors.Wrapf(err, "failed to refresh bootstrap token")
+	}
+	// NB: this may not be sufficient to keep the token live if we don't see it before it expires, but when we generate a config we will set the status to "ready" which should generate an update event
+	return ctrl.Result{
+		RequeueAfter: DefaultTokenTTL / 2,
+	}, nil
+}
+
+func (r *KubeadmConfigReconciler) rotateMachinePoolBootstrapToken(ctx context.Context, config *bootstrapv1.KubeadmConfig, cluster *clusterv1.Cluster, scope *Scope) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Config is owned by a MachinePool, checking if token should be rotated")
+	remoteClient, err := r.remoteClientGetter(ctx, r.Client, util.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	token := config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token
+	shouldRotate, err := shouldRotate(ctx, remoteClient, token)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if shouldRotate {
+		log.V(2).Info("creating new bootstrap token")
+		token, err := createToken(ctx, remoteClient)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to create new bootstrap token")
+		}
+
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken.Token = token
+		log.Info("Altering JoinConfiguration.Discovery.BootstrapToken", "Token", token)
+
+		// update the bootstrap data
+		return r.joinWorker(ctx, scope)
+	}
+	return ctrl.Result{
+		RequeueAfter: DefaultTokenTTL / 3,
+	}, nil
```

Comment:
Could we switch this return and the one from the function above to be `ctrl.Result{}, error` instead?

Reply:
How do I make sure it gets reqeued before the token expiration with `ctrl.Result{}` ?

Reply:
@vincepri not sure how to address this